python/ray/air/checkpoint.py

```diff
@@ -230,6 +232,22 @@ def _metadata(self) -> _CheckpointMetadata:
             },
         )
 
+    def _copy_metadata_attrs_from(self, source: "Checkpoint") -> None:
+        """Copy in-place metadata attributes from ``source`` to self."""
+        for attr, value in source._metadata.checkpoint_state.items():
+            if attr in self._SERIALIZED_ATTRS:
+                setattr(self, attr, value)
+
+    @_metadata.setter
+    def _metadata(self, metadata: _CheckpointMetadata):
+        if metadata.checkpoint_type is not self.__class__:
+            raise ValueError(
+                f"Checkpoint type in metadata must match {self.__class__}, "
+                f"got {metadata.checkpoint_type}"
+            )
+        for attr, value in metadata.checkpoint_state.items():
+            setattr(self, attr, value)
```

Comment:
Should this be abstracted into a separate function? Not sure if a setter should have side effects.

Reply:
I suppose the entire purpose of this setter is to set attributes :) If we decide to store metadata in a single dict instead of using attributes, it will be easy to change here.